Add general `partial_implementation` guideline

ddbeck opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->

#### Summary

This PR provides a general guideline for setting `partial_implementation`.

<!-- ✍️ In a sentence or two, describe your changes. -->

#### Related issues

This continues a discussion started in https://github.com/mdn/browser-compat-data/pull/26605#issuecomment-2838278224.

<!-- 🔨 If applicable, use "Fixes #XYZ" -->

<!-- ✅ After submitting, review the results of the "Checks" tab! -->

ddbeck commented:
I suppose one thing that is potentially missing here—or at least not directly mentioned—is some notion of severity: how do we know to use a note alone versus note+partial? A good example of this being a concern is on something like Safari's top-level await bug (https://github.com/mdn/browser-compat-data/pull/26510) or `css.properties.accent-color.maintains_contrast` (https://github.com/mdn/browser-compat-data/pull/26605). This guideline basically says that severity is no matter; if it's inconsistent and difficult to detect as such, we flag it as partial.

My inclination is to adopt a guideline like this, which doesn't require a judgment call on severity, and wait to see if we find a situation where the guideline requires us to do something overtly silly.

caugner commented on docs/data-guidelines/index.md:
```diff
@@ -52,6 +52,30 @@ Do not use `"preview"` for planned but not yet implemented support changes. In o
+
+- `"partial_implementation": true`: One browser implements `ExampleInterface` but exposes it on the `Node` interface instead of the specified `Element` interface.
+  Other implementers use the correct parent interface.
+  This misleads feature detection (`"ExampleInterface" in Element.prototype` → false), so it's partially implemented.
```
Nit: The word "misleads" keeps confusing me here. Would this be right:

```suggestion
  This fails feature detection (`"ExampleInterface" in Element.prototype` → false), so it's partially implemented.
```

ddbeck replied:
See my previous comment about the interpretation of "failure" in this setting.

We might make this more obvious with a different example though. Perhaps something like:

> `"partial_implementation": true`: One browser exposes a method, `example()`, but due to a bug it's a no-op. Other browsers implement the method's intended behavior. This confuses feature detection because `"example" in ExpectedInterface.prototype` returns `true`, even though the method does not work.

caugner replied:
Yes, I like that a lot, although I would suggest replacing "it's a no-op" with something like "it has no observable behavior or effect" to make it clearer.

ddbeck replied:
I did this, somewhat modified, in 16aab1545ac411d42b683deae50db657f909ff10. It's actually a real example with the names changed: [`Notification()` in Chrome for Android](https://developer.mozilla.org/en-US/docs/Web/API/Notification/Notification#browser_compatibility).

caugner reviewed (commented):
LGTM so far, just one nit.

And one possible extension: What about BCD features with subfeatures? In general, we don't set `partial_implementation: true` on the parent feature, e.g. a CSS property, if the information can be captured on a subfeature, e.g. a CSS property value with `partial_implementation: true` or `version_added: false`.

caugner commented on docs/data-guidelines/index.md:
```diff
@@ -52,6 +52,30 @@ Do not use `"preview"` for planned but not yet implemented support changes. In o
+
+- The browser's support does not implement mandatory specified behavior.
+- The browser's support is inconsistent with at least one other browser.
+- The browser's support misleads or evades feature detection.
```
Dito (unless there is an important difference between "failing" vs "misleading" and "evading":

```suggestion
- The browser's support fails feature detection.
```

ddbeck replied:
I interpret "failing" feature detection to be a normal event ("feature detection failed, therefore this browser doesn't support the feature, now falling back to something else…") versus the thing I'm trying to capture here, where the browser produces a confusing result (i.e., showing support when it doesn't actually support it).

Maybe something like this might do?

```suggestion
- The browser's support confuses, evades, or deceives feature detection.
```

caugner replied:
This goes in the right direction. As a non-native English speaker, the nuances between confusing, evading, and deceiving are not obvious to me though.

Is the gist of this equivalent to the following:

```suggestion
- The browser's support cannot be easily or accurately identified using feature detection.
```

Also, is this condition independent, or does it actually refer to the previous two conditions. And if so, would the following version be accurate:

```suggestion
- The missing or inconsistent behavior cannot be easily or accurately identified using feature detection.
```

ddbeck replied:
I've got another idea for you. What do you think of reducing the number of terms? I don't think the nuance of the other terms is critical. Mostly I was trying to be comprehensive (and to shortcircuit some rules lawyering from vendors, "but we _meant_ to do it this way, so it's fine").

```suggestion
- The browser's support causes confusing feature detection results.
```

To respond to your suggestion, "easily or accurately" doesn't work for me—I think it's a big change from my original proposal. For one, "easy" is subjective. I doubt we could agree on what it means among BCD maintainers, much less with developers in general.

As for "accurately," what I'm trying to get at here is that if a developer makes _any_ reasonable attempt to detect if a feature is available, but the browser _still_ doesn't behave as expected then that's a `partial_implementation`. I don't care that there might exist _some_ accurate check. Developers shouldn't have to run a comprehensive test suite (in production, on every page load!) to find out if a feature is available. And as a maintainer, I shouldn't have to say to developers, "well, you didn't check in precisely the right way to find out that this feature isn't _really_ supported" to justify our `partial_implementation` data.

caugner replied:
Thanks, I think your latest suggestion is sufficiently clear.

ddbeck commented:
> And one possible extension: What about BCD features with subfeatures? In general, we don't set `partial_implementation: true` on the parent feature, e.g. a CSS property, if the information can be captured on a subfeature, e.g. a CSS property value with `partial_implementation: true` or `version_added: false`.

If you wish to make this explicit you can, though we don't really document (for example) promoting `"version_added": false` to parent features. Though a possible coda to this guideline might be:

> Setting `partial_implementation` stands alone. Unlike `"version_added" false`, `partial_implementation` does not dictate support data to descendant features. Likewise, a subfeature's `"version_added": false` never justifies `"version_added": false` or `"partial_implementation": true` for an ancestor feature.

caugner commented:
> Though a possible coda to this guideline might be:
> 
> > Setting `partial_implementation` stands alone. Unlike `"version_added": false`, `partial_implementation` does not dictate support data to descendant features. Likewise, a subfeature's `"version_added": false` never justifies `"version_added": false` or `"partial_implementation": true` for an ancestor feature.

Yes, please, I like that.

caugner commented on docs/data-guidelines/index.md:
```diff
@@ -52,6 +52,33 @@ Do not use `"preview"` for planned but not yet implemented support changes. In o
+Unlike `"version_added": false`, `partial_implementation` does not dictate support data to descendant features.
+Likewise, a subfeature's `"version_added": false` does not imply `"version_added": false` or `"partial_implementation": true` for an ancestor feature.
+
+Here are some example situations:
+
+- `"partial_implementation": false`: All implementing browsers ignore part of a feature's specified behavior in the same way.
+  This behavior is consistent and is a _de facto_ complete implementation.
+
+- `"partial_implementation": false`: All implementing browsers provide a form control user interface, but the specification gives the implementer discretion over its look and feel.
+  A developer complains that one browser's user interface lacks some desired quality that other browsers implement; they want it to be marked as partially implemented.
+  Use a note or non-standard behavioral subfeature instead.
```
```suggestion
**Don't set** `partial_implementation` when all implementing browsers …

1. … ignore part of a feature's specified behavior in the same way
  (this consistent behavior constitutes a _de facto_ complete implementation), or
2. … provide a form control user interface, but the specification gives the implementer discretion over its look and feel
  (if one browsers' user interface lacks some quality that is desired by developers, and that other browsers implement, create a non-standard behavioral subfeature instead).
```

caugner commented on docs/data-guidelines/index.md:
```diff
@@ -52,6 +52,33 @@ Do not use `"preview"` for planned but not yet implemented support changes. In o
+  Use a note or non-standard behavioral subfeature instead.
+  <!-- TODO: link to behavioral subfeature guidelines, when available -->
+
+- `"partial_implementation": true`: `CSS.supports()` returns `true` for a property name and value, but the value has no behavior.
+  See [Non-functional defined names imply `"partial_implementation"`](#non-functional-defined-names-imply-partial_implementation).
+
+- `"partial_implementation": true`: One browser exposes a constructor, `Example()`, but it always throws an error. Other browsers implement the constructor's intended behavior. This confuses feature detection because `typeof Example === "function"` returns `true`, even though the constructor does not work.
```
```suggestion
Set `partial_implementation: true` in these cases:

- `CSS.supports()` returns `true` for a property name and value, but the value has no behavior in this browser, and the value has the intended behavior in another browser.
  _See also_: [Non-functional defined names imply `"partial_implementation"`](#non-functional-defined-names-imply-partial_implementation).

- The browser exposes a constructor, `Example()`, but it always throws an error, and another browser implements the constructor's intended behavior.
  _Note_: This confuses feature detection because `typeof Example === "function"` returns `true`, even though the constructor does not work.
```

pygy commented:
As mentioned in https://github.com/mdn/browser-compat-data/issues/26838#issuecomment-2895858454, I think that point 1...

> 1. The browser's support does not implement mandatory specified behavior.

... should be sufficient to mark support as partial if the missing behavior pertains to accessibility, especially if the lack of support isn't obvious to the layman as the feature seems to work.

ddbeck commented:
@pygy For context, this PR is meant to regularize a process we already do here, so I don't think introducing new criteria is in scope. But for completeness, I'll say that I don't believe that deviation from a specification _alone_ is a workable trigger for `partial_implementation`. Specifications evolve well ahead of implementations, often in highly speculative ways that no vendor has committed to implementing. If we relied solely on spec compliance then many interoperable features would need to be marked as partially implemented despite having no meaningful compatibility risks to developers.

There is a concurrent proposal, https://github.com/mdn/browser-compat-data/pull/26781, to regularize "behavioral subfeatures" which typically capture the kind of relatively anonymous specification evolution that has gone on for #26838.

That said, nothing in this PR (or the behavioral subfeature PR) should be understood to forbid (non-partial) notes about bugs and other limitations, accessibility or otherwise.

captainbrosset commented on docs/data-guidelines/index.md:
```diff
@@ -52,6 +52,33 @@ Do not use `"preview"` for planned but not yet implemented support changes. In o
+You must set `"partial_implementation": true` when all of the following conditions are met:
+
+- The browser's support does not implement mandatory specified behavior.
+- The browser's support is inconsistent with at least one other browser.
```
What if it's the other browser that's wrong?

caugner replied:
I think what's meant here is:

> At least one other browser implements this behavior.

captainbrosset replied:
Ok. In that case, I'd suggest rephrasing.
Right now, this reads as "set partial implementation to true if this browser doesn't work like that other browser". And to me, I have no way of know which browser is right.

hamishwillee replied:
@caugner (as you know and have commented) Given an interface, it is not uncommon for new properties or methods to be added later. At that point the parent interface becomes partial even if it was complete previously. My understanding though is that you wouldn't "go back" and capture that in the interface level. The interface would remain complete?

chrisdavidmills replied:
My understanding of "complete" would be "complete at the time the data was written", so it would still hold true if another feature was added later. Does that need clarifying? It would be a real mess if people started going around retroactively marking data as partial because new features were added later.

ddbeck replied:
To answer some questions here:

> Right now, this reads as "set partial implementation to true if this browser doesn't work like that other browser". And to me, I have no way of know which browser is right.

All three conditions must be met. The _spec_ is almost always right—unless all implementers ignore the spec same exact way. It's possible to have partials for _all_ implementers, if they all differ from the spec _and_ they're not all mutually compatible.

> At that point the parent interface becomes partial even if it was complete previously. My understanding though is that you wouldn't "go back" and capture that in the interface level. The interface would remain complete?

Partial data doesn't cascade to descendants and it doesn't bubble up to parents. [`Intl`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#browser_compatibility) is probably the idealized case: `Intl` is fully supported ever since it was exposed because being exposed is all it does. The fact that `Intl.Segmenter` was added much later doesn't matter for `Intl`—we don't retroactively mark it as partial.

I'm not trying to propose any radical changes here. Instead, I'm trying to describe what we usually do most of the time anyway.

caugner commented:
[Here is an interesting case](https://github.com/mdn/browser-compat-data/pull/26993#discussion_r2183223360) from WebDriver BiDi: The [`browsingContext.print()` command](https://w3c.github.io/webdriver-bidi/#command-browsingContext-print) takes a parameter `page` (also in BCD) with two fields `height` and `width`, and the spec defines the minimum size as 1x1, and the following:

> If the implementation is unable to provide a paginated representation of navigable for any reason then return [error](https://w3c.github.io/webdriver/#errors) with [error code](https://w3c.github.io/webdriver/#dfn-error-code) [unsupported operation](https://w3c.github.io/webdriver/#dfn-unsupported-operation).

Before Firefox 133, a size of 0x0 was accepted, and did not result in an `unsupported operation` error.

I don't think it makes sense to set partial implementation here, even though strictly speaking this behavior
- is mandatory and specified,
- is inconsistent with another browser (Chrome), and
- confuses feature detection results.

Am I missing something?

ddbeck commented:
@caugner that is an interesting example! I have a few thoughts on this:

First, in an earlier draft (before I submitted the PR), I had a caveat about how the partial implementation must correspond to "reasonable developer behavior."

I didn't want to allow situations where a contrived scenario would lead to a partial implementation when it would have no real-world meaning. I dropped it from the draft because I didn't quite know how to define "reasonable" but this is the kind of scenario I was thinking of. It doesn't much matter that Firefox and Chrome are incompatible because there's no compelling reason for a developer to expect producing a non-conforming dimensionless PDF to work.

Second, looking into the both the bug and specification text seems to be in response to a challenge when it comes to testing the API (read https://github.com/w3c/webdriver-bidi/issues/473). It seems like the sort of thing that we would have never known about except that:

- WPT revealed the incompatibility.
- The specifiers resolved the matter by revising the specification by narrowing allowed behavior, instead of changing the test or allowing implementers discretion (search the issue for "for any reason").
- The behavior's "defect" was known to BCD only because the vendor proactively provided the information.

At no point in this story was an ordinary web developer involved. This suggests to me that there's the possibility for a fourth condition to be added to the list:

>  The browser's support inconsistency has a demonstrable real-world negative impact on web developers.

I would take such a condition to mean that there has to be a bug reported (or replicated) by a developer in a situation that impacts real-world users (including preventing the feature from being used with real-world users). Or, in the case of a feature that has not yet shipped (e.g., in a beta), the authors and reviewers of the PR have good reason to believe that it will have that effect when it does ship.

ddbeck commented:
This came up again in the BCD call today. I've applied a few changes:

- Added a first attempt at a developer significance requirement here as "The browser's support has a demonstrable negative impact on web developers." with a new example where a partial wouldn't be set (a failing WPT with no developer complaints).
- Added "feedback welcome" text at the top of the guideline. In today's call, we talked about adopting the guideline with the understanding it's sort of provisional. We talked about removing this text in 6 months (assuming the guideline survives that long).
- Added a link to this guideline from the docs index.

@caugner and @Elchi3, I'd welcome re-review here. Thank you!

ddbeck commented:
From today's BCD call: this still needs a link to this PR, for symmetry with other guidelines. I'll try to get to this shortly.
